acpi: build the RSDP and pass it to the guest

dlrobertson opened the pull request:
Build the RSDP and pass it to the guest via the seabios table-loader feature.
We also now fill out the MADT with a implementation of the `SDTBuilder`
trait and the `MADTBuilder`

As seen below, we do indeed control the RSDP and the Interrupt Control
Address (in this case a bad address).

```
GUEST0: Copying ACPI RSDP from 0x40000008 to 0x000f4ca0
GUEST1: Copying ACPI RSDP from 0x40000008 to 0x000f4ca0
...
GUEST0: [    0.000000][    T0] ACPI: RSDP 0x00000000000F4CA0 000024 (v02       )
GUEST1: [    0.000000][    T0] ACPI: RSDP 0x00000000000F4CA0 000024 (v02       )
GUEST0: [    0.000000][    T0] ACPI: XSDT 0x0000000040001000 00002C (v01                 00000000      00000000)
GUEST1: [    0.000000][    T0] ACPI: XSDT 0x0000000040001000 00002C (v01                 00000000      00000000)
GUEST0: [    0.000000][    T0] ACPI: APIC 0x0000000040002000 000040 (v05                 00000000      00000000)
GUEST1: [    0.000000][    T0] ACPI: APIC 0x0000000040002000 000040 (v05                 00000000      00000000)
GUEST0: [    0.000000][    T0] ACPI: Local APIC address 0xdeadbeef
GUEST1: [    0.000000][    T0] ACPI: Local APIC address 0xdeadbeef
```

Fixes: #69

ALSchwalm commented:
Rather than patching seabios, I think the correct way to do this is via [qemu_fw_cfg](https://github.com/mythril-hypervisor/seabios/blob/4fd52bd592255c7c1a422f0eead10ef2279643ed/src/fw/paravirt.c#L532). Ultimately we should write a table-loader that seabios will use [here](https://github.com/mythril-hypervisor/seabios/blob/4fd52bd592255c7c1a422f0eead10ef2279643ed/src/fw/paravirt.c#L189), but we may just be able to use one from qemu (like we do for most of the linux-dma loading).

ALSchwalm commented on mythril_core/src/vm.rs:
```diff
@@ -16,6 +16,9 @@ use alloc::sync::Arc;
 use alloc::vec::Vec;
 use spin::RwLock;
 
+pub type ImageMapper =
+    dyn Fn(&[u8], &GuestPhysAddr, &mut GuestAddressSpace) -> Result<()>;
+
```
Not even a comment? For shame! But seriously, I'm not clear what the idea is here? Is this trying to express that it is a function that maps some collection of bytes to a location in the guest address space? It seems like the implementation on `VirtualMachine` in `map_data` is pretty much the only reasonable way to do that. I'd rather just make that a method on `GuestAddressSpace`.

dlrobertson replied:
😆  yeah I should have added a comment. I think a method on the `GuestAddressSpace` would be better. I'll add that in the update.

ALSchwalm reviewed (changes requested):
This is looking pretty good. I'm not sure about the merge order of this vs https://github.com/mythril-hypervisor/mythril/pull/82

dlrobertson commented:
Updated to use the seabios table-loader feature instead of the seabios patch

ALSchwalm commented on .gitmodules:
```diff
@@ -2,6 +2,7 @@
 	path = seabios
 	url = https://github.com/mythril-hypervisor/seabios.git
 	shallow = true
+	branch = acpi-next
```
We don't need this anymore, right?

dlrobertson replied:
Yeah not needed.

ALSchwalm commented on mythril/src/acpi/madt.rs:
```diff
@@ -251,6 +253,44 @@ impl Ics {
     }
 
+    /// Encode into the byte sequence
+    pub fn encode(&self) -> Result<Vec<u8>> {
```
I'd really rather we structure this stuff so that we are given a buffer and write into it. Ultimately I'd like to move away from mythril requiring Alloc, so we should avoid using Vec wherever possible

dlrobertson replied:
Yeah I can work on that.

ALSchwalm commented on mythril/src/acpi/rsdt.rs:
```diff
@@ -229,3 +234,157 @@ impl<'a> Iterator for RSDTIterator<'a> {
+    }
+
+    /// Add the given System Descriptor Table to the RSDT.
+    pub(super) fn add_sdt<T: SDTBuilder>(&mut self, builder: T) -> Result<()> {
```
I would rather this be written `builder: impl SDTBuilder`

dlrobertson commented on mythril/src/acpi/seabios.rs:
```diff
@@ -0,0 +1,175 @@
+
+impl<'a> TryInto<Vec<u8>> for TableLoaderCommand<'a> {
+    type Error = Error;
+    fn try_into(self) -> Result<Vec<u8>> {
```
I can move this to an encode implementation that takes a buffer too.

dlrobertson commented on mythril/src/acpi/rsdt.rs:
```diff
@@ -229,3 +234,157 @@ impl<'a> Iterator for RSDTIterator<'a> {
+        table_loader: &mut TableLoaderBuilder,
+    ) -> Result<()> {
+        let table_length = self.map.iter().count() * 8;
+        let xsdt = build_sdt(b"XSDT", 0, 0, &vec![0x00; table_length])?;
```
We don't know how big this buffer will be, and we don't know how big the entire table loader buffer will be, so I'm not sure the best way to move away from alloc is, but there are some options.

ALSchwalm commented on mythril/src/acpi/rsdt.rs:
```diff
@@ -229,3 +234,157 @@ impl<'a> Iterator for RSDTIterator<'a> {
+    /// The identifying signature for this table type.
+    const SIGNATURE: [u8; 4];
+
+    /// The revision of the tabe
```
Typo here and the next function

ALSchwalm commented on mythril/src/acpi/seabios.rs:
```diff
@@ -0,0 +1,175 @@
+use crate::error::{Error, Result};
```
I'd like all new modules going forward to require docs for everything, so we can eventually just require it for the whole project. It would probably also be good to write a bit of a module description given that you had to kind of reverse-engineer this from seabios, right?

dlrobertson replied:
Yeah I'll add module level docs for that with the protocol described. Everything in the acpi is `deny(missing_docs)`, so everything in here should be documented, but some documentation should probably be improved (like the module level docs).

ALSchwalm commented on scripts/seabios.config:
```diff
@@ -11,6 +11,7 @@ CONFIG_QEMU=y
 # CONFIG_CSM is not set
 CONFIG_QEMU_HARDWARE=y
 # CONFIG_XEN is not set
+CONFIG_MYTHRIL=y
```
I assume this can also be dropped

ALSchwalm commented:
While we will definitely need something like your `RingBuffer` at some point (so we can have a lock free producer consumer for inter-vm communication and such)., I'm not really sure why we need it here? Why not just use arrayvec, which we already depend on?

dlrobertson commented:
Good question. I've thought the same thing and debated the use of a ring for the table loader builder. It definitely shouldn't be used in the SDT building like I currently have things. I ended up using it because 1) the table loader builder will eventually support multiple "loads" of commands, where the pending commands will be loaded (no explicit need of a ring buffer here, but it seemed like a natural fit) 2) something like it will be useful in the future especially as we move away from `alloc` 3) I thought about it and felt like writing it :smile: I'll definitely remove it from the SDT builder and can remove it from the table loader builder if you'd rather stick to `arrayvec`. The ring buffer stuff might be best added in a separate effort.

ALSchwalm commented:
Gotcha, yeah I'd rather it come in separately, but def push a branch so we
have it available

On Wed, Dec 9, 2020, 10:03 PM Dan Robertson <notifications@github.com>
wrote:

> Good question. I've thought the same thing and debated the use of a ring
> for the table loader builder. It definitely shouldn't be used in the SDT
> building like I currently have things. I ended up using it because 1) the
> table loader builder will eventually support multiple "loads" of commands,
> where the pending commands will be loaded (no explicit need of a ring
> buffer here, but it seemed like a natural fit) 2) something like it will be
> useful in the future especially as we move away from alloc 3) I thought
> about it and felt like writing it 😄 I'll definitely remove it from the
> SDT builder and can remove it from the table loader builder if you'd rather
> stick to arrayvec. The ring buffer stuff might be best added in a
> separate effort.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mythril-hypervisor/mythril/pull/76#issuecomment-742224452>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIDYPYTCY6LYMRWVG2UU5DSUBCB3ANCNFSM4PMPJAOQ>
> .
>

dlrobertson commented:
The `acpi` mod should have have no usage of `Vec` and should use `ArrayVec` instead. Switched everything including the table loader away from the ring buffer too and removed it.
